lib/polly.ml

```diff
@@ -70,11 +75,17 @@ let create = caml_polly_create1
 
 let close t = Unix.close t
 
-let add = caml_polly_add
+let add t fd e =
+  let r = caml_polly_add t fd e in
+  if r < 0 then caml_raise_unix_error "Polly.add"
 
-let del t fd = caml_polly_del t fd Events.empty
+let del t fd =
+  let r = caml_polly_del t fd Events.empty in
+  if r < 0 then caml_raise_unix_error "Polly.del"
```

Comment:
If we need to preserve `errno` it should be returned by the C function. With multiple threads there is no guarantee that it is still valid for `call_raise_unix_error` to pick it up.

Reply:
The code may work correctly as it is currently written, but it relies on some fragile properties.

 `errno` is typically per-thread (if it wasn't then there would be no safe way to read it in a multi-threaded program).
But if any code would be executed inbetween return from `epoll_ctl` and reading the errno inside the other C stub then `errno` may have already been overwritten. This could happen if a signal is delivered to the program, which would be executed the next time OCaml checks for pending signals. The way the code is currently written it doesn't allocate between the call the `caml_polly_...` and the call to `caml_raise...`, but that may break easily if the code is changed in the future (e.g. someone adds a debugging statement).
Newer version of OCaml may use an `[@poll error]` attribute to detect these cases, but it'd be simpler and more robust to do as suggested and return `-errno` on error.